py/dml/c_backend.py

```diff
@@ -2210,88 +2240,66 @@ def register_saved_attributes(initcode, node):
     initcode.out('}\n', preindent=-1)
     initcode.out('}\n', preindent=-1)
 
-def generate_alloc_trait_vtables(node):
-    start_function_definition('void _allocate_traits(void)')
-    out('{\n', postindent=1)
-    for subnode in flatten_object_subtree(node):
-        if subnode.traits:
-            for trait in subnode.traits.referenced:
-                name = subnode.traits.vtable_cname(trait)
-                out(f'{name} = MM_MALLOC(1, typeof(*{name}));\n')
-    out('}\n', preindent=-1)
-    splitting_point()
-
 def generate_init_trait_vtables(node, param_values):
     '''Return a list of pairs (dimsizes, 'call();\n')
     meaning that call() is to be invoked with variables (_i0, _i1, ..) given by
     all possible indices within dimsizes'''
     generate_vtable_instances(node)
     for trait in traits.Trait.referenced:
         generate_tinit(trait)
-    for subnode in flatten_object_subtree(node):
+    all_nodes = list(flatten_object_subtree(node))
+    for subnode in all_nodes:
         generate_trait_trampolines(subnode)
-    init_calls = init_trait_vtables_for_node(node, param_values, (), ())
-    by_dims = {}
-    for (dims, call) in init_calls:
-        by_dims.setdefault(dims, []).append(call)
-    all_dims = sorted(by_dims)
-    fun_count = 0
-    call_count = 0
-    def finish_current_function(last_dims):
-        for _ in last_dims:
-            out('}\n', preindent=-1)
+    # Hack: Split up trait initialization in small functions, to avoid
+    # performance problems with `gcc -fvar-tracking`. This is
+    # required because of a GCC bug that is fixed in GCC 12:
+    # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104468
```

Comment:
Is this logic tested?

Reply:
It is smoke-tested in the sense that we have lots of tests where the split happens. We don't test that it solves a performance problem, but that would be hard to do.